### ISSUE TITLE
Issues reading and writing evolved records with the old schemas

### ISSUE THREAD

**Issue by julianpeeters** [julianpeeters/avro-scala-macro-annotations#17]
Say we want to add a field to a record, then write more records the old schema. 

In Java this is possible by using the new record with the old schema, as long as you used that exact schema for reading (or else use the new schema provided it has default values that can fill in for the lack of the added field).  

But something goes wrong when using Java from this library. 

Reference case class:
```scala
 Case Class MyRecord(i: Int)`
 val rec = MyRecord(0)
```
`rec` gets encoded as a `0` in the byte array.

"Evolved" case class (the nullable field `j` has been added):
```scala
Case Class MyRecord(i: Int, j: Option[Int])
val rec = MyRecord(0, None)
```
`rec` gets encoded as 2 bytes: one to represent the value of the int, the other to specify the first member of the array (in this case a null, so no other bytes are written).

Writing a new, "evolved" record with the old schema seems to work correctly, as `rec` appears to be converted into a `0` in the byte array. Just like it did before the new field was added correctly, just like it did when Java handled this correctly.

But what works in Java fails when reading. We get `{"i": 0,"j":1}`instead of `{"i": 0}`. A default value for the int within the option slipped in.

Will add a fix and some tests.

More complete example:
```scala
package com.example

import com.julianpeeters.avro.annotations._

import org.apache.avro.specific.SpecificDatumWriter
import org.apache.avro.specific.SpecificDatumReader
import org.apache.avro.io.{DecoderFactory, EncoderFactory}

import java.io.ByteArrayOutputStream


@AvroRecord
case class MyRecord(var i: Int, var j: Option[Int])

object Main {

  def main(args: Array[String]) = {
     val myRecord = MyRecord(i=1, j=null)
     val subSchema = new org.apache.avro.Schema.Parser().parse("""{"type":"record","name":"MyRecord","namespace":"com.twc.needle.domain","doc":"Auto-generated schema","fields":[{"name":"i","type":"int","doc":"Auto-Generated Field"}]}
""")
   val writer = new SpecificDatumWriter[MyRecord](subSchema)

    val out = new java.io.ByteArrayOutputStream()
    val encoder = EncoderFactory.get().binaryEncoder(out, null)
    writer.write(myRecord, encoder)


    encoder.flush
    val ba = out.toByteArray
    out.close

    val reader = new SpecificDatumReader[MyRecord](subSchema)
    val decoder = DecoderFactory.get().binaryDecoder(ba, null)
    val decoded = reader.read(null, decoder)


    println(decoded.toString)
  }
}
```

**Comment by julianpeeters** [comment]
Mostly a false alarm.

Not only is this is a misuse of Avro (the record class' fields and the writer's schema must evolve in lockstep), there was also an error in the schema of my test that caused the datum to fall through to Generic.  When the test is fixed, the mismatch between the incorrect use in Avro and this library becomes {"i": 0,"j":1} instead of {"i": 0,"j":null}. Still, it would be nice if even incorrect usage behaved the same across Avro and this library, which is just a matter of using a smarter default value for the no-args constructor of the record's class.

**Comment by julianpeeters** [comment]
Fixed in 0.8.0
